lib/mayaUsd/render/vp2RenderDelegate/mesh.cpp

```diff
@@ -1832,118 +1871,145 @@ void HdVP2Mesh::_UpdateDrawItem(
-                    }
+        // If the enable state is changed, then update it.
+        if (stateToCommit._enabled != nullptr) {
+            renderItem->enable(*stateToCommit._enabled);
+        }
+
+        ProxyRenderDelegate& drawScene = param->GetDrawScene();
+
+        // TODO: this is now including all buffers for the requirements of all
+        // the render items on this rprim. We could filter it down based on the
+        // requirements of the shader.
+        if (stateToCommit._geometryDirty || stateToCommit._boundingBox) {
+            MHWRender::MVertexBufferArray vertexBuffers;
+
+            for (auto& entry : *primvarInfo) {
+                const TfToken&            primvarName = entry.first;
+                MHWRender::MVertexBuffer* primvarBuffer = nullptr;
+                if (isBBoxItem && primvarName == HdTokens->points) {
+                    primvarBuffer
+                        = const_cast<MHWRender::MVertexBuffer*>(sharedBBoxGeom.GetPositionBuffer());
                 } else {
-                    drawScene.setInstanceTransformArray(
-                        *renderItem, stateToCommit._instanceTransforms);
+                    primvarBuffer = entry.second->_buffer.get();
+                }
+                if (primvarBuffer) { // this filters out the separate color & alpha entries
+                    vertexBuffers.addBuffer(primvarName.GetText(), primvarBuffer);
                 }
+            }
+
+            // The API call does three things:
+            // - Associate geometric buffers with the render item.
+            // - Update bounding box.
+            // - Trigger consolidation/instancing update.
+            drawScene.setGeometryForRenderItem(
+                *renderItem, vertexBuffers, *indexBuffer, stateToCommit._boundingBox);
+        }
 
-                if (stateToCommit._instanceColors.length()
-                    == newInstanceCount * kNumColorChannels) {
+        // Important, update instance transforms after setting geometry on render items!
+        auto& oldInstanceCount = stateToCommit._renderItemData._instanceCount;
+        auto  newInstanceCount = stateToCommit._instanceTransforms.length();
+
+        // GPU instancing has been enabled. We cannot switch to consolidation
+        // without recreating render item, so we keep using GPU instancing.
+        if (stateToCommit._renderItemData._usingInstancedDraw) {
+            if (oldInstanceCount == newInstanceCount) {
+                for (unsigned int i = 0; i < newInstanceCount; i++) {
+                    // VP2 defines instance ID of the first instance to be 1.
+                    drawScene.updateInstanceTransform(
+                        *renderItem, i + 1, stateToCommit._instanceTransforms[i]);
+                }
+            } else {
+                drawScene.setInstanceTransformArray(*renderItem, stateToCommit._instanceTransforms);
+            }
+
+            // upload any extra instance data
```

Comment:
This is the real difference in this code block. I make additional calls to MPxSubSceneOverride::setExtraInstanceData().

I don't have support for generic additional instanced streams because the USD name of the stream is not necessarily the name of the stream in the shader. If we had some way to handle that cleanly then we wouldn't have to make display color a special case.

Reply:
I guess there is maybe a performance optimization here where we update the values one by one. @HdC-adsk do you know if calling that method in a loop is faster then updating the whole buffer in one shot? I see the code for updating transforms does it one by one if possible.

Reply:
Actually the two branches have been there before I joined the USD project. Maybe @kxl-adsk remembers the history?

My guess is that if in the future we can have fine-grain update for instance transforms (e.g. one moving instance with all the other being static), the loop could be a better option. For now I don't know how it can be faster.